rust/arrow/src/ipc/reader.rs

```diff
@@ -651,19 +654,20 @@ impl<R: Read + Seek> RecordBatchReader for FileReader<R> {
         self.schema.clone()
     }
 
-    fn next_batch(&mut self) -> Result<Option<RecordBatch>> {
```

Comment:
Is this change (to not need `mut`) required in this PR? It seems like it could be separated out from the DataFusion change, though perhaps I am missing something

Given `self.reader` and `self.current_block` are actually being changed by `next_batch` I think it might be harder to reason about if this method looks like it doesn't mutate any state. 

Also, once we make this non-mut, if we ever wanted to make it mut in the future, that would likely be a compatibility breaking change.

Reply:
The issue is that both DataFusion and Parquet implement the same RecordBatchReader trait that requires `&mut self` and therefore requires the use of mutexes. Maybe DataFusion should have its own trait instead. I'll try that approach next.